### PR TITLE
Fix instruments not filtered by method in Worksheet Template edit view

### PR DESCRIPTION
**Pull request by Lunga001**
## Description of the issue/feature this PR addresses

Linked issue: https://github.com/senaite/senaite.core/issues/

## Current behavior before PR
On the worksheet template on edit mode, selecting a method does not filter the instrument that belong to selected method
## Desired behavior after PR is merged
 Method selection should filter the instrument by the selected method
 
--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

**Comment by ramonski** [comment]
Hi @Lunga001, what is the state of this PR?

**Comment by Lunga001** [comment]
Updated @ramonski the PR is ready. Sorry about the delay, just that I wasn't sure removing selection widget was the best way to go about it.

File: src/bika/lims/content/worksheettemplate.py [hunk 1]
```diff
@@ -100,16 +99,6 @@
...
-            format="select",
-            label=_("Method"),
-            description=_(
-                "Restrict the available analysis services and instruments"
-                "to those with the selected method."
-                " In order to apply this change to the services list, you "
-                "should save the change first."
-            ),
-        ),
     ),
```
Comment (xispa) [inline src/bika/lims/content/worksheettemplate.py]: Removing this widget makes the filtering of instruments when a method is selected to not work at all. The reason is that the javascript in charge of updating the list of methods expects the control to be a `SelectionWidget` here: https://github.com/senaite/senaite.core/blob/2.x/src/senaite/core/browser/static/js/bika.lims.worksheettemplate.js#L11

File: src/bika/lims/browser/worksheettemplate.py [hunk 2]
```diff
@@ -31,16 +29,20 @@ class AJAXGetWorksheetTemplateInstruments(BrowserView):
...
+        if 'method_uid' not in self.request.keys():
             return json.dumps(items)
+
+        method_uid = str(json.loads(self.request.get('method_uid', '')))
+        if not method_uid:
+            return json.dumps(items)
+
+        try:
+            instruments = api.get_object(method_uid).getInstruments()
+        except api.APIError:
```
Comment (xispa) [inline src/bika/lims/browser/worksheettemplate.py]: Better remove this try/except. System might fail silently otherwise.

**Review by xispa — CHANGES_REQUESTED** [review]
Thanks @Lunga001, but it does not work, see comments. 

Note this `AJAXGetWorksheetTemplateInstruments` class and the javascript are old-school. You might consider to use the "EditForm" introduced with https://github.com/senaite/senaite.core/pull/1719 to achieve this instrument-filtering functionality in a more elegant way, and with extrapoints for the removal of old code.

**Comment by Lunga001** [comment]
> Thanks @Lunga001, but it does not work, see comments.
> 
> Note this `AJAXGetWorksheetTemplateInstruments` class and the javascript are old-school. You might consider to use the "EditForm" introduced with #1719 to achieve this instrument-filtering functionality in a more elegant way, and with extrapoints for the removal of old code.

Gave EditForm a shot

File: src/senaite/core/browser/form/adapters/configure.zcml [hunk 3]
```diff
@@ -32,6 +32,20 @@
...
+      factory=".worksheettemplate.EditForm"/>
+  <!--
+  <adapter
+      name="validate_instruments"
+      for="bika.lims.interfaces.IWorksheetTemplate
+           bika.lims.interfaces.IMethod
+           zope.publisher.interfaces.browser.IBrowserRequest"
+      factory=".method.ValidateInstrumentsAdapterBase"/>
+  -->
+
```
Comment (xispa) [inline src/senaite/core/browser/form/adapters/configure.zcml:40]: Please remove this instead if commenting if is not used

File: src/senaite/core/browser/form/adapters/worksheettemplate.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    def modified(self, data):
+        name = data.get("name")
+        value = data.get("value")
+
+        # Handle Methods Change
+        if name == "RestrictToMethod" and value:
+            # Available Methods
+            empty = [{"title": _("No Instrument"), "value": [""]}]
+            # Get selected method
+            method = api.get_object_by_uid(value[0])
```
Comment (xispa) [inline src/senaite/core/browser/form/adapters/worksheettemplate.py]: Note that `value[0]` can be empty. You get `APIError: get_object_by_uid requires UID as first argument; got  instead` in such case. Remember you can check if something is an UID with `api.is_uid(someting)`

File: src/senaite/core/browser/form/adapters/worksheettemplate.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            instruments = self.get_available_instruments_for(method)
```
Comment (xispa) [inline src/senaite/core/browser/form/adapters/worksheettemplate.py]: You can omit this `get_available_instruments_for` and rely on method's `getInstruments` directly here, cause in accordance with my previous comment, at this point you should have a valid method already

File: src/senaite/core/browser/form/adapters/worksheettemplate.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+                    "options": i_opts})
+                self.add_update_field("Instruments:list", {
+                    "options": i_opts, "selected": i_sel})
+                self.add_update_field("Instrument", {
+                    "options": empty + i_opts})
+            else:
+                self.add_update_field("Instruments:list", {
+                    "options": []})
+                self.add_update_field("Instruments_options", {
+                    "options": i_opts})
```
Comment (xispa) [inline src/senaite/core/browser/form/adapters/worksheettemplate.py]: this should be empty instead of `i_opts`, right?

**Review by xispa — CHANGES_REQUESTED** [review]
@Lunga001 please pull, note I've pushed a commit for cleanup here https://github.com/senaite/senaite.core/pull/1887/commits/3312335b1fd072b02fc375e8cadc16a33ec55125

I also left some comments

**Review by xispa — APPROVED** [review]

**Pull request merged**